File: app/src/main/java/se/curtrune/lucy/classes/calender/CalenderDate.java
package se.curtrune.lucy.classes.calender;

import androidx.annotation.NonNull;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import se.curtrune.lucy.classes.item.Item;

public class CalenderDate {
    private LocalDate date;
    private List<Item> items = new ArrayList<>();

    public LocalDate getDate(){
        return date;
    }
    public int getDay(){
        return date.getDayOfMonth();
    }
    public List<Item> getItems() {
        return items;
    }
    public boolean hasEvents() {
        return items.size() > 0;
    }
    public void setDate(LocalDate date) {
        this.date = date;
    }
    public void setItems(List<Item> items){
        this.items = items;
    }
    public CalenderDate(){

    }
    public CalenderDate(LocalDate date){
        this.date = date;
    }
    public CalenderDate(LocalDate date, Item item){
        this.date = date;
        this.items.add(item);
    }
    public CalenderDate(LocalDate date, List<Item> items){
        this.date = date;
        this.items = items;

    }

    @NonNull
    @Override
    public String toString() {
        return String.format(Locale.getDefault(), "%s, %d events", date.toString(), items.size());
    }
    public DayOfWeek getFirstDayOfMonth(){
        LocalDate firstDate = date.withDayOfMonth(1);
        return firstDate.getDayOfWeek();
    }
    public void addItems(List<Item> items) {
        this.items.addAll(items);
    }

    public void add(Item item) {
        if(items == null){
            items = new ArrayList<>();
        }
        items.add(item);
        //TODO sort items
    }
}

File: app/src/main/java/se/curtrune/lucy/classes/calender/CalenderDate.kt
package se.curtrune.lucy.classes.calender

import se.curtrune.lucy.classes.Mental
import se.curtrune.lucy.classes.item.Item
import se.curtrune.lucy.workers.MentalWorker
import java.time.DayOfWeek
import java.time.LocalDate
import java.util.Locale

class CalenderDate {
    @JvmField
    var date: LocalDate = LocalDate.now()
    //var mental: Mental = Mental()
    //TODO private set or something
    var items: MutableList<Item> = ArrayList()

    val day: Int
        get() = date.dayOfMonth

    fun hasEvents(): Boolean {
        return items.isNotEmpty()
    }

/*    fun setItems(items: MutableList<Item>) {
        this.items = items
    }*/

    constructor()
    constructor(date: LocalDate) {
        this.date = date
    }

    constructor(date: LocalDate, item: Item) {
        this.date = date
        items.add(item)
    }

    constructor(date: LocalDate, items: MutableList<Item>) {
        this.date = date
        this.items = items
    }

    override fun toString(): String {
        return String.format(Locale.getDefault(), "%s, %d events", date.toString(), items.size)
    }

    val firstDayOfMonth: DayOfWeek
        get() {
            val firstDate = date.withDayOfMonth(1)
            return firstDate.dayOfWeek
        }

    fun addItems(items: List<Item>) {
        this.items.addAll(items)
    }

    fun add(item: Item) {
        items.add(item)
        //TODO sort items
    }
    val events: List<Item>
        get() = items.filter { it.isCalenderItem }.sortedBy { it.targetTimeSecondOfDay }
    val mental: Mental
        get() = MentalWorker.getMental(items)
}
